form: revamp errors/warning in feedback

ptamarit opened the pull request:
* Fixes inveniosoftware/react-invenio-forms#277

> - Updates FormFeedback messages and structure to show relevant sections and count, with directed links to them
> - Doesn't stop publishing in case of severity checks
> - Updates files upload button to have error class

<img width="1654" alt="Screenshot 2025-03-13 at 09 52 16" src="https://github.com/user-attachments/assets/99488fae-b223-448a-af77-691f87704ebe" />

<img width="1467" alt="Screenshot 2025-03-13 at 16 23 27" src="https://github.com/user-attachments/assets/81a4c612-d0dc-4a15-8b46-981ba79c0269" />

<img width="1586" alt="Screenshot 2025-03-13 at 16 29 07" src="https://github.com/user-attachments/assets/3d70d880-5657-4f8b-8d2a-6711d67677ec" />

0einstein0 commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/errors/FormFeedback.js:
```diff
@@ -134,122 +112,77 @@ class DisconnectedFormFeedback extends Component {
-      let messages = labelledErrorMessages[label] || [];
-      labelledErrorMessages[label] = messages.concat(step1[key]);
-    }
+    const orderedSections = [
```
Need a better way of ordering the sections.

0einstein0 commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/api/DepositRecordSerializer.js:
```diff
@@ -369,7 +369,14 @@ export class RDMDepositRecordSerializer extends DepositRecordSerializer {
     //                 enough for now.
     for (const e of errors) {
-      _set(deserializedErrors, e.field, e.messages.join(" "));
+      if ("severity" in e && e.severity !== "error") {
```
remove severity !== error check

kpsherva commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/errors/FormFeedback.js:
```diff
@@ -134,122 +111,76 @@ class DisconnectedFormFeedback extends Component {
+      }
+
+      // If not matched in predefined sections, check dynamically in accordions
+      const sectionElement = document
```
this regex match looks a little bit risky, why do we need it?

0einstein0 replied:
This makes it possible to match the label, e.g. for="metadata.creators" instead of trying to match a specific label for "metadata.creators.0" or "metadata.creators.0.identifiers.scheme" . 
 `<label for="metadata.creators" class="field-label-class invenio-field-label"><i aria-hidden="true" class="user icon"></i>Creators</label>`

kpsherva commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/errors/FormFeedback.js:
```diff
@@ -134,122 +111,76 @@ class DisconnectedFormFeedback extends Component {
-    }
+    const orderedSections = [
+      ...accordions.filter((id) => errorSections.has(id)), // Keep sections that exist in order
+      ...[...errorSections.keys()].filter((id) => !accordions.includes(id)), // Append missing sections
```
in which cases a section is considered missing?

0einstein0 replied:
it orders the sections based on the accordions order found in the dom structure, but some sections for example visibility is not under an accordion but rather a card so it appends them at the end. 

<img width="888" alt="Screenshot 2025-03-17 at 20 28 33" src="https://github.com/user-attachments/assets/3ff11871-396e-4b59-8c63-d14b57e0ad54" />

kpsherva commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/errors/FormFeedback.js:
```diff
@@ -134,122 +111,76 @@ class DisconnectedFormFeedback extends Component {
+    return orderedSections.map((sectionId, i) => {
+      const { label, count } = errorSections.get(sectionId);
+      return (
+        <a key={i} className="pl-5" href={`#${sectionId}`}>
```
please fix the eslint warning.
the reason why it flags it, is that using iterators in arrays is sometimes messing with the rendering of specific elements - it does not know which ones are really updated as components. Can you use section IDs as keys instead?

kpsherva commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/errors/FormFeedback.js:
```diff
@@ -134,122 +111,76 @@ class DisconnectedFormFeedback extends Component {
     const errors = errorsProp || {};
 
-    const { feedback, message } = _get(ACTIONS, actionState, {
+    let { feedback, message } = _get(ACTIONS, actionState, {
```
why `let` instead of `const`?

0einstein0 replied:
It was reassigning the feedback later in the code

kpsherva commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/errors/FormFeedback.js:
```diff
@@ -275,12 +212,30 @@ class DisconnectedFormFeedback extends Component {
         positive={feedback === "positive"}
         warning={feedback === "warning"}
         negative={feedback === "negative"}
+        info={feedback === "suggestive"}
```
nit:
```suggestion
         const feedback = {};
         feedback[`${feedback}`] = True
        <Message {...feedback} ...
```

kpsherva commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/CreatibutorsField/CreatibutorsField.js:
```diff
@@ -73,25 +72,36 @@ class CreatibutorsFieldForm extends Component {
     const creatibutorsError =
       error || (creatibutorsList === formikInitialValues && initialError);
 
+    let className = "";
```
I remember we discussed this couple of times, from my understanding you can also do only
```suggestion
    const className = contributors.severity;
```
because you also pass errors in the new format

kpsherva commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/CreatibutorsField/CreatibutorsField.js:
```diff
@@ -73,25 +72,36 @@ class CreatibutorsFieldForm extends Component {
+        typeof creatibutorsError !== "string" ? creatibutorsError.severity : "error";
+    }
+
+    let generalCreatibutorsErrorMessage;
+    if (typeof creatibutorsError === "string") {
+      generalCreatibutorsErrorMessage = creatibutorsError;
+    } else if (typeof creatibutorsError === "object") {
+      generalCreatibutorsErrorMessage = creatibutorsError?.message;
+    }
```
I think this one needs a comment on why we do this and what is "general" error
also to improve readability:
```suggestion
    let generalCreatibutorsErrorMessage = contributorsError;
    if (typeof creatibutorsError === "object") {
      generalCreatibutorsErrorMessage = creatibutorsError?.message;
    }
```

kpsherva commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/CreatibutorsField/CreatibutorsField.js:
```diff
@@ -73,25 +72,36 @@ class CreatibutorsFieldForm extends Component {
-                  identifiersError={identifiersError}
+                  creatibutorError={
+                    creatibutorsError &&
+                    typeof creatibutorsError !== "string" &&
```
```suggestion
                    objectErrorFormat &&
```

kpsherva commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/fields/License/LicenseField.js:
```diff
@@ -73,9 +73,21 @@ class LicenseFieldForm extends Component {
+    const initialError = getIn(initialErrors, fieldPath, null);
+    const licenseError = error || (licenseList === formikInitialValues && initialError);
+
+    let className = "";
+    if (licenseError) {
+      className = typeof licenseError !== "string" ? licenseError.severity : "error";
+    }
```
see my previous comment on contributors:
```suggestion
    let className = license.severity;
```

kpsherva commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/state/actions/deposit.js:
```diff
@@ -63,10 +63,34 @@ export const saveDraftWithUrlUpdate = async (draft, draftsService) => {
   return response;
 };
 
+function _hasValidationErrorsWithSeverityError(errors) {
+  if (typeof errors === "object") {
+    if (Object.hasOwn(errors, "message") && Object.hasOwn(errors, "severity")) {
+      if (errors["severity"] === "error") {
+        return true;
+      }
+    }
+    for (const key of Object.keys(errors)) {
+      if (key !== "message" && key !== "severity") {
+        return _hasValidationErrorsWithSeverityError(errors[key]);
+      }
+    }
+  } else {
+    // If the error message is a string and not an object with `message` and `severity` keys, then it's an error.
+    return true;
+  }
+}
+
```
for readability:
```suggestion
function _hasValidationErrorsWithSeverityError(errors) {
  const isObjectError = typeof errors === "object";
  const isSeverityError = errors["severity"] === "error";
  
  if (isObjectError && isSeverityError) {
        return true;
    }
    for (const key of Object.keys(errors)) {
    \\ we need an explanation here on why we do this, I am not understanding why :)
      if (key !== "message" && key !== "severity") {
        return _hasValidationErrorsWithSeverityError(errors[key]);
      }
    }
    
   // If the error message is a string and not an object with `message` and `severity` keys, then it's an error.
   return true;
}

```

i removed:
```
if (Object.hasOwn(errors, "message") && Object.hasOwn(errors, "severity")) 
```
because this can obfuscate the fact that we are getting some other object than we expect as the error. In this case we should fail, not hide it :)
also, it helps with readability

0einstein0 commented on invenio_rdm_records/assets/semantic-ui/js/invenio_rdm_records/src/deposit/errors/FormFeedback.js:
```diff
@@ -134,122 +111,76 @@ class DisconnectedFormFeedback extends Component {
+        ?.closest(".accordion");
+      if (sectionElement) {
+        const sectionId = sectionElement.id;
+        const label = sectionElement.getAttribute("label") || "Unknown section";
```
It can be unknown when the closest .accordion element related to an error doesn't have a label attribute.

ptamarit commented:
Unaddressed comments related to the fact that we mix current error format with new error format.
To be addressed once the new format is used everywhere.
